cpp/include/raft/linalg/cholesky_r1_update.cuh

```diff
@@ -132,6 +132,137 @@ void choleskyRank1Update(const raft::handle_t& handle,
+ *                                       &n_bytes, CUBLAS_FILL_MODE_LOWER,
+ *                                       stream);
+ * rmm::device_uvector<char> workspace(n_bytes, stream);
+ *
+ * for (n=1; n<=n_rows; rank++) {
+ *   // Calculate a new row/column of matrix A into A_new
+ *   // ...
+ *   // Copy new row to L[rank-1,:]
+ *   RAFT_CUBLAS_TRY(cublasCopy(handle.get_cublas_handle(), n - 1, A_new, 1,
+ *                           L + n - 1, ld_L, stream));
+ *   // Update Cholesky factorization
+ *   raft::linalg::choleskyRank1Update(
+ *       handle, L, rank, ld_L, workspace, &n_bytes, CUBLAS_FILL_MODE_LOWER,
+ *       stream);
+ * }
+ * Now L stores the Cholesky decomposition of A: A = L * L.T
+ * @endcode
+ *
+ * - Upper triangular factorization:
+ * @code{.cpp}
+ * // Initialize arrays
+ * int ld_U = n_rows;
+ * rmm::device_uvector<math_t> U(ld_U * n_rows, stream);
+ * raft::linalg::choleskyRank1Update(handle, L, n_rows, ld_U, nullptr,
+ *                                       &n_bytes, CUBLAS_FILL_MODE_UPPER,
+ *                                       stream);
+ * rmm::device_uvector<char> workspace(stream, n_bytes, stream);
+ *
+ * for (n=1; n<=n_rows; n++) {
+ *   // Calculate a new row/column of matrix A into array A_new
+ *   // ...
+ *   // Copy new row to U[:,n-1] (column major layout)
+ *   raft::copy(U + ld_U * (n-1), A_new, n-1, stream);
+ *   //
+ *   // Update Cholesky factorization
+ *   raft::linalg::choleskyRank1Update(
+ *       handle, U, n, ld_U, workspace, &n_bytes, CUBLAS_FILL_MODE_UPPER,
+ *       stream);
+ * }
+ * // Now U stores the Cholesky decomposition of A: A = U.T * U
+ * @endcode
+ *
+ * @tparam ElementType The data-type of raft::mdspan
+ * @tparam IndexType Integer used for addressing
+ * @param handle RAFT handle (used to retrive cuBLAS handles).
+ * @param L raft::device_matrix_view<ElementType, raft::col_major> to store the
+ *        triangular matrix L, and the new column of A in column major format, size [n*n]
+ * @param n number of elements in the new row.
+ * @param ld stride of columns in L
+ * @param workspace optional raft::device_vector_view<char> shall be std::nullopt ar an array
+ *    of size [n_bytes].
+ * @param n_bytes size of workspace is returned here if workspace==std::nullopt.
+ * @param uplo indicates whether L is stored as an upper or lower triangular
+ *    matrix (CUBLAS_FILL_MODE_UPPER or CUBLAS_FILL_MODE_LOWER)
+ * @param eps numerical parameter that can act as a regularizer for ill
+ *    conditioned systems. Negative values mean no regularizaton.
+ */
+template <typename ElementType, typename IndexType = std::uint32_t>
+void cholesky_rank1_update(const raft::handle_t& handle,
+                           raft::device_matrix_view<ElementType, IndexType, raft::col_major> L,
+                           int n,
+                           int ld,
+                           std::optional<raft::device_vector_view<char>> workspace,
```

Comment:
I'd actually prefer for us to not expose the workspace at all. This is something we can add back if needed, but we spent a lot of time supporting it in the early APIs and it turns out it would just get wrapped and no longer used directly.

Reply:
How would we not expose the workspace without exposing an interface to accept an allocator?

Reply:
Perhaps we could hide a scratch space allocation in the RAFT handle.  That would ensure coupling of the allocation to a CUDA stream, permitting e.g., asynchronous allocations.

Reply:
Take a similar case in the `raft::distance` API as an example: https://github.com/rapidsai/raft/blob/0d71ceb613432a4cd438b78a9180666e411f64c4/cpp/include/raft/distance/distance.cuh#L215.

Reply:
I don't know if this is possible to use, because this is an iterative method that is called from the host multiple times by the user. The workspace is used repeatedly for every iteration. So is it okay if I keep the current design?

Reply:
I would prefer we not keep workspaces on the public API. As discussed in the past, a pool allocator will remove the overhead from intermediate allocations if that becomes a bottleneck here. Something tells me it's not the bottleneck, though. Another option we have, which I was recently discussing w/ @mhoemmen (and I've discussed w/ others in the past) is to add an option in the handle to attach an additional allocator just for intermediate / temporary allocations in between calls to primitives (this allocator would be beneficial as a pool allocator).

Reply:
That would be a future discussion, would it not? Allocating and de-allocating a workspace every iteration will severely slow this function down though if a synchronous allocator is used

Reply:
The more I think about this one, the more I'm realizing this might be an indicator of a bad code smell. This doesn't really fit the patterns used in the other primitives and really what it means is that the LARS solver itself should be in RAFT (which is the the plan) and this primitive should really be an implementation detail. 

Maybe we should skip mdspanifying this function altogether and leave a note in the original code that it's not moving to the new API.

Reply:
>  Allocating and de-allocating a workspace every iteration will severely slow this function down though if a synchronous allocator is used

The idea behind attaching an optional allocator to the handle is that a pool allocator could be used for temporary workspace allocations in the primitives while a non-pool allocator (could be device synchronous, asynchronous, or stream synchronous, really) could be used everywhere else globally. That type of design would keep the knowledge of workspaces out of the public APIs while still giving the control over the allocations to the user.

Reply:
Sounds good, I can remove the API then